Adding stronger types to Cotter interfaces

theFong opened the pull request:
Im not sure how the Cotter.d.ts is generated but I have added stronger types in hopes of strengthening the interface.

alecfong committed:
adding stronger types to Cotter interfaces

theFong commented:
I would like to add stronger types to the `verificationProccessPromise` but looks like an interface for the verifySuccess payload has not been defined yet?

alecfong committed:
added payload interface based on experimentation and added types promise resolution

theFong commented:
I added a payload interface to type `verificationProccessPromise` resolution. Probably could be used in other publicly exposed parts of the api. Let me know if you would like the naming changed etc or want it moved somewhere else :)

alecfong committed:
adding payload to api -- making some assumptions

theFong commented on src/API.ts:
```diff
@@ -138,7 +139,7 @@ class API {
     credential: any,
     origin: string
-  ): Promise<any> {
+  ): Promise<Payload> {
```
Assuming this is the same interface

theFong commented on src/API.ts:
```diff
@@ -211,7 +212,7 @@ class API {
     origin: string,
     publicKey: string
-  ): Promise<any> {
+  ): Promise<Payload> {
```
Assuming this is the same interface

putrikarunia commented:
Hey, thank you for contributing! I think verifySuccess is supposed to be the response that is returned to the client because it was used for `resolve(self.verifySuccess)`. Is it different when you tried it? 

It should be like this:

```typescript
export interface VerifySuccess {
  email?: String;
  phone?: String;
  oauth_token: OAuthToken;
  user: User;
  identifier?: String; // only from WebAuthn same as email/phone
  status?: String; // only from WebAuthn = SUCCESS or CANCELED
}
```
where `OAuthToken` is from `handler/TokenHandler.ts` but missing some parts:
```diff

export interface OAuthToken {
  access_token: string;
  id_token: string;
  token_type: string;
  refresh_token: string;
+  expires_in: number;
+  auth_method: string;
}
```
and `User` from `models/User`.

alecfong committed:
making changes to pass payload in to onbegins as well as adding proper verifysuccess interface

theFong commented:
> Hey, thank you for contributing! I think verifySuccess is supposed to be the response that is returned to the client because it was used for `resolve(self.verifySuccess)`. Is it different when you tried it?
> 
> It should be like this:
> 
> ```ts
> export interface VerifySuccess {
>   email?: String;
>   phone?: String;
>   oauth_token: OAuthToken;
>   user: User;
>   identifier?: String; // only from WebAuthn same as email/phone
>   status?: String; // only from WebAuthn = SUCCESS or CANCELED
> }
> ```
> 
> where `OAuthToken` is from `handler/TokenHandler.ts` but missing some parts:
> 
> ```diff
> export interface OAuthToken {
>   access_token: string;
>   id_token: string;
>   token_type: string;
>   refresh_token: string;
> +  expires_in: number;
> +  auth_method: string;
> }
> ```
> 
> and `User` from `models/User`.

Oops! You are totally correct the payload interface was something I was logging in the onBegin and not on the verifySuccess. I have added those accordingly.
Added the VerifySuccess interface to binder.ts. Is this the proper file to use?

alecfong committed:
adding styles interface based on https://docs.cotter.app/sdk-reference/web/web-sdk-verify-email-phone/styling and adding more complete definitons to the config

alecfong committed:
fixing up some interfacing and resolving errors in cotter

alecfong committed:
using wrong interface

alecfong committed:
addressing errors

putrikarunia commented on src/binder.ts:
```diff
@@ -76,3 +83,46 @@ export interface Config {
+  redirect_url: string;
+  signature: string;
+  timestamp: string;
+}
```
the Payload that is passed into `OnBegin` should be like this
```typescript
export interface Payload {
  auth_required: boolean;
  device_name: string; // User Agent
  device_type: "BROWSER";
  identifier: string; // User's email/phone
  identifier_type: "EMAIL" | "PHONE";
}
```
What you're seeing above (challenge, challenge_id, etc) is probably from some other console log which I should remove from the code.

putrikarunia reviewed (changes requested):
Looks good 👍, just need to update the `Payload` to the one I posted in the comment.

alecfong committed:
making payload reflect accurate object + name change

theFong commented:
Changed Payload to reflect real object and changed name from `Payload` to `OnBeginPayload`
